types/react/index.d.ts

```diff
@@ -264,6 +265,25 @@ declare namespace React {
         props?: Q, // should be Q & Attributes
         ...children: ReactNode[]): ReactElement<P>;
 
+    // Context via RenderProps
+    type Provider<T> = ComponentType<{
+        value: T,
+        children?: ReactNode
+    }>;
+    type Consumer<T> = ComponentType<{
```

Comment:
it may be worthwhile to refactor the props interface of `Consumer` and `Provider` in case those props are spread from another place

Reply:
@ferdaber I'm not sure I understand the specific refactoring you want here. Could you elaborate?

Reply:
I'm talking about refactoring the props as their own exportable interface like this:
```tsx
interface ConsumerProps<T> {
  children(value: T): ReactNode,
  unstable_observedBits?: number
}
interface ProviderProps<T> {
  children?: ReactNode
  // if you choose to use my suggestion, otherwise just T
  value: T extends undefined ? never : T 
}

// this way a consumer of the type can import `ConsumerProps` and `ProviderProps` for other purposes
type Consumer<T> = ComponentType<ConsumerProps<T>>
type Provider<T> = ComponentType<ProviderProps<T>>
```

Reply:
Hmmm.. When I try that, in the test code proposed below it no longer is able to determine the type of `value`:

```
const ContextRenderPropsConsumerComponent = (
    <ContextWithDefaultValue.Consumer>
        {value => value.assert} // <--- Value here is now "any" with the props refactor
    </ContextWithDefaultValue.Consumer>
);
```

See the test example below for the full proposed test code.

Reply:
@ferdaber It looks like the refactoring you proposed might cause some issues as well, listed in the comment above. Do you see a way around them?

Reply:
odd... I tested it out and it works fine for me, these are my relevant snippets (I copied your test code as-is):
```tsx
// react/index.d.ts

    interface ProviderProps<T> {
        value: T;
        children?: ReactNode;
    }

    interface ConsumerProps<T> {
        children: (value: T) => ReactNode;
        unstable_observedBits?: number;
    }

    type Provider<T> = ComponentType<ProviderProps<T>>;
    type Consumer<T> = ComponentType<ConsumerProps<T>>;
    interface Context<T> {
        Provider: Provider<T>;
        Consumer: Consumer<T>;
    }
    function createContext<T>(
        defaultValue: T,
        calculateChangedBits?: (prev: T, next: T) => number
    ): Context<T>;
    function createContext<T>(): Context<T | undefined>;
```
```tsx
// react/test/tsx.tsx

import React from '.'

const RenderPropsContext = React.createContext('defaultValue');

interface RenderPropsContextTest {
    assert: boolean;
}

const ContextWithDefaultValue = React.createContext<RenderPropsContextTest>({ assert: true });
const ContextWithoutDefaultValue = React.createContext<RenderPropsContextTest>();

const ContextRenderPropsConsumerComponent = (
    <ContextWithDefaultValue.Consumer>
        {value => value.assert}
    </ContextWithDefaultValue.Consumer>
);

const ContextRenderPropsProviderConsumerComponent = (
    <ContextWithoutDefaultValue.Provider value={{ assert: true }}>
        <ContextWithoutDefaultValue.Consumer>
            {value => value.assert}
        </ContextWithoutDefaultValue.Consumer>
    </ContextWithoutDefaultValue.Provider>
); // $ExpectError
```

Reply:
@scally wrote:
> When I try that, in the test code proposed below it no longer is able to determine the type of value:

perhaps that was an IDE bug? There are cases where Webstorm & IntelliJ IDEA don't show inferred argument type correctly (claims it is `any`), but when used incorrectly it still emits error with message mentioning correct type of the argument